src/walkTree.ts

```diff
@@ -2,53 +2,89 @@ import { isObject } from './isObject';
+ * Recursively walks tree, copy values to revision object and cloning
+ * paths that have cloned.
  */
 export const walkTree = <T>(source: T, revision: T): T => {
-  // Return early for no op
+  // Return early for no op.
   if (source === revision) {
     return source;
   }
 
-  // Check if both objects are the same. This catches Arr/Obj merging
+  // Return revision if its a non object type.
+  // This also catches null/undefined values.
+  if (!isObject(source) || !isObject(revision)) {
+    return revision;
+  }
+
+  // Check if both objects are the same. This catches Arr/Obj merging.
   const matchingType = source instanceof revision.constructor &&
                        revision instanceof source.constructor;
 
   if (!matchingType) {
     return revision;
   }
 
-  // Return revision if its a non object type
-  if (!isObject(source) || !isObject(revision)) {
-    return revision;
-  }
-
-  // Gather keys. For arrays, its just a list of indexes
-  // TODO (asif) See if it'll be quicker walk arrays instead of getting its keys
+  // Gather keys. For arrays, its just a list of indexes.
+  // TODO(asif): Test walking arrays instead of getting its keys.
   type keyType = keyof T;
   const sourceKeys = Object.keys(source) as keyType[];
-  // const revisionKeys = Object.keys(revision) as keyType[];
-  let changed = false;
+  const revisionKeys = Object.keys(revision) as keyType[];
+
+  let cloned = false;
+  let deletedCount = 0;
 
   for (const key of sourceKeys) {
     const sourceValue = source[key];
     const revisionValue = revision[key];
+
+    // Recursive call. If newValue is deeply equal to sourceValue, then 
+    // we return sourceValue, allowing us to skip cloning.
     const newValue = walkTree(sourceValue, revisionValue);
 
-    // If walkTree doesn't return new object, that means no change
+    // If walkTree doesn't return new object, that means no change.
     if (newValue === sourceValue) {
       continue;
     }
 
     // On first change, clone current object.
-    // This needs to be done before we set the newValue to the property
-    // TODO(asif): Remove isObject call when TS/pull/13288 is merged in
-    if (!changed && isObject(source)) {
+    // This needs to be done before we set the newValue to the property, so 
+    // we don't modify the original object.
+    // TODO(asif): Remove isObject call when TS/pull/13288 is merged in.
+    if (!cloned && isObject(source)) {
       source = shallowClone(source);
-      changed = true;
+      cloned = true;
     }
 
-    source[key] = newValue;
+    if (newValue === undefined) {
+      deletedCount++;
+      // TODO(asif): Investigate performance of delete vs setting undefined. 
+      // TODO(asif): Add configuration property to prefer undefined vs delete.
+      delete source[key];
+    } else {
+      source[key] = newValue;
+    }
+  }
+
+  // Check if there are any new keys in revision object.
+  if (sourceKeys.length + deletedCount === revisionKeys.length) {
```

Comment:
How does this work?  From my reading, sourceKeys is the keys of the old object, and revisionKeys is the keys of the new object, and deletedCount is how many we removed.

Wouldn't something like have 3 keys and remove 1 (adding none) lead to 3 + 1 === 2 and miss the check?

Reply:
Good catch. this should be sourceKeys.length - deletedCount.

I need to add a unit test for this; any ideas?